### PR TITLE
Add asset upload functionality

### PR DESCRIPTION
**Pull request by ra0x3**
### Description
- This PR looks large but it isn't -- most of the diff is just moving code around  (only 1K new lines)
  - Further about half of those added 1K lines are just queries 🥲 
- This PR does a few things:
  - Manually downgrades `anyhow` back to `1.53`
  - Converts the `database` package into the `fuel-indexer-database` crate
     - Crate reservations (some of which we may not need) are [here](https://github.com/FuelLabs/crate-reservations/pull/10)
  - Moves `IndexerConfig` logic from `fuel-indexer-lib/lib.rs` to `fuel-indexer/config.rs`
  - Moves the database components  (`IndexerConnection`, `IndexerConnectionPool`, and the queries) from `fuel-indexer-schema/{db, models}.rs` to `fuel-indexer-database/{lib, queries}.rs`
    - Still a little more to move from `fuel-indexer-schema` into `fuel-indexer-database` (e.g., the database model structs, `ForiengKey`) but that can wait
  - Adds new migrations for `index_asset_registry_*` tables which will hold the uploaded assets
    - ⚠️ **You'll need to check these tables when testing**
  - Introduces an `"identifier"` to an index -- with the `identifier` just being the name of the index
    - The idea is that an index can be globally identified as `$namesapce + $identifier`
  - Adds placeholder authentication middleware
    - Will need to add some concept of assets "belonging to" an operator (i.e., address/wallet) before this can actually be used
  - Allows users to upload assets (manifest, schema, wasm)
  - Updates all relevant docs and examples

### Some implementation details

⚠️ **Testing steps are pretty painful/extensive -- happy to hop on a call and pair through this**

- Assets each kept in separate tables (schema, wasm, manifest)
- When new assets are uploaded to the same index, versions are bumped (just a simple integer count increased by one)
- An index uses the latest version of its assets
- You can either upload all assets at once, or upload them individually
- Service currently requires a restart after asset upload
  - With our EKS setup this isn't a problem
  - However, obviously the ideal case is not having to restart the service at all
    - Might require some logic to pass some state to the `IndexerService` 👀 
- If asset blob checksums/digests are the same (e.g., you're uploading a copy of the same blob) then no data is inserted
- If no multipart is in the request -- you get a 405
- If you include a multipart field that isn't schema, wasm, manifest -- you get a 405
- No sqlx-data.json was generated from `cargo sqlx prepare --database-url sqlite:///path-to/my.db` 🤷 

### Testing steps
- [ ] Stop your local postgres if you have one running
- [ ] Spin up the end-to-end test `bash tests/e2e/composable-indexer/compose-up.bash`
- [ ] Register a set of assets...(below command)

```bash
curl -v http://0.0.0.0:29987/api/index/counter/index1 \
    -F "manifest=@examples/counter/manifest.yaml" \
    -F "wasm=@target/wasm32-unknown-unknown/release/counter_indexer.wasm" \
    -F "schema=@examples/counter/schema/counter.graphql" \
    -H 'Content-type: multipart/form-data' -H "Authorization: foo" | json_pp
```
- [ ] Check the relevant database tables to ensure your assets were uploaded
- [ ] Try running that same `curl` command again -- you should still only see a single set of assets (duplicates aren't uploaded)
- [ ] Now try to register a second set of assets (below command)

```bash
curl -v http://0.0.0.0:29987/api/index/composability_test/index1 \
    -F "manifest=@tests/e2e/composable-indexer/manifest.yaml" \
    -F "wasm=@target/wasm32-unknown-unknown/release/composable_indexer_lib.wasm" \
    -F "schema=@tests/e2e/composable-indexer/composable-indexer-lib/schema/schema.graphql" \
    -H 'Content-type: multipart/form-data' -H "Authorization: foo" | json_pp
```
- [ ] You should see these assets uploaded successfully to a separate index
- [ ] Now make some change to `examples/counter/manifest.yaml` then re-upload it using the below command

```bash
curl -v http://0.0.0.0:29987/api/index/counter/index1 \
    -F "manifest=@examples/counter/manifest.yaml" \
    -H 'Content-type: multipart/form-data' -H "Authorization: foo" | json_pp
```
- [ ] You should see that you now have two assets for the `counter.index1` index, and the latest version in `index_asset_registry_manifest` should be "2"
- [ ] Stop the `fuel-indexer` Docker container 
- [ ] Start the `fuel-indexer` container again
- [ ] Making a request to `/ping` should still work (because the indices were loaded from the database when the service was started again)

### CODE REVIEW

File: fuel-indexer/src/manifest.rs [hunk 1]
```diff
@@ -37,13 +38,24 @@ impl Manifest {
         Ok(manifest)
     }
 
+    pub fn to_bytes(&self) -> Vec<u8> {
+        serde_yaml::to_string(&self)
+            .expect("Failed here")
```
Comment (deekerno) [inline fuel-indexer/src/manifest.rs:43]: Could we put a more descriptive message here?

File: fuel-indexer/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+#[derive(Deserialize)]
+pub struct TmpIndexerConfig {
+    pub fuel_node: Option<FuelNodeConfig>,
+    pub graphql_api: Option<GraphQLConfig>,
+    pub database: Option<DatabaseConfig>,
+}
+
+impl IndexerConfig {
+    // FIXME
```
Comment (deekerno) [inline fuel-indexer/src/config.rs:372]: Is this still needed? If so, can we put a more descriptive message here?

**Review by deekerno — CHANGES_REQUESTED** [review]

**Comment by ra0x3** [comment]
@deekerno 
- Agree with all the feedback so far. 
- If there's no other feedback after a few days I'll merge as is and follow up in a small PR (so as to minimize the review noise).
- If there is other feedback I'll just make the changes in this PR

**Review by deekerno — APPROVED** [review]

File: fuel-indexer/src/api.rs [hunk 3]
```diff
@@ -106,6 +108,126 @@ pub async fn health_check(
...
+    Path((namespace, identifier)): Path<(String, String)>,
+    Extension(schema_manager): Extension<Arc<RwLock<SchemaManager>>>,
+    Extension(pool): Extension<IndexerConnectionPool>,
+    multipart: Option<Multipart>,
+) -> (StatusCode, Json<Value>) {
+    if let Some(mut multipart) = multipart {
+        let mut conn = pool
+            .acquire()
+            .await
+            .expect("Failed to get database connection.");
```
[the remark this answers is not in the document]
Comment (ra0x3) [reply fuel-indexer/src/api.rs:148]: https://github.com/FuelLabs/fuel-indexer/issues/215

File: fuel-indexer/src/api.rs [hunk 4]
```diff
@@ -126,23 +248,30 @@ impl GraphQlApi {
...
         }
 
         let graph_route = Router::new()
-            .route("/:name", post(query_graph))
+            .route("/:namespace", post(query_graph))
+            .layer(Extension(schema_manager.clone()))
+            .layer(Extension(pool.clone()));
+
+        let asset_route = Router::new()
+            .route("/:namespace/:identifier", post(register_index_assets))
```
[the remark this answers is not in the document]
Comment (ra0x3) [reply fuel-indexer/src/api.rs:260]: @tjsharp1 
- Yea that's the idea. `namespace` is your org, `identifier` is the name of that index running for that org
- You're right we'll eventually want friendly messaging around what namespace's are available
  - We would first need to implement some actual form of user authentication so namespaces can be tied to authenticated users

**Review by deekerno — APPROVED** [review]

**Pull request merged**